### ISSUE TITLE
Serialization API

### ISSUE THREAD

**Issue by kvark** [slide-rs/specs#18]

**Comment by White-Oak** [comment]
Woah, this is gonna be interesting!

**Comment by Aceeri** [comment]
Would it be alright for `specs` to primarily use `serde` for serialiation purposes?

**Comment by kvark** [comment]
Yes, behind a feature gate.

**Comment by Aceeri** [comment]
So I figured out a way to have it abstracted a bit:
```rust
pub trait Deserializer<T> {                               
    /// Deserialize data into an entity.                  
    fn deserialize(&mut World, T) -> Result<(), Error>;   
                                                          
    /// Called when the world has a component registered. 
    fn register_component<C>(&mut self) { }               
                                                          
    /// Called when the world has a resource registered.  
    fn register_resource<R>(&mut self) { }       

    /// Maybe some other methods related to the world's state being changed?         
}                                                         
                                                          
pub trait Serializer<T> {                                 
    /// Serialize a part of the world and return it.      
    fn serialize(&mut World) -> Result<T, Error>;         
                                                          
    /// Called when the world has a component registered. 
    fn register_component<C>(&mut self) { }               
                                                          
    /// Called when the world has a resource registered.  
    fn register_resource<R>(&mut self) { }                
}                                                         
```

However, I've run into a bit of a wall. In the files you need some way to identify the component as itself, which would probably be some sort of built in static id for each `Component` type. 
e.g.
.yaml
```yaml
---
Component1:
    field1: 5
    field2: false

Component2:
    field3: null
```

.json
```json
{
    "Component1": {
        "field1": 5,
        "field2": false
    },
    "Component2": {
        "field3": [ 1, 1, 1 ]
    }
}
```

So I'm thinking of maybe having the `Component` trait include another function: `fn id() -> String` or something similar. This would probably be nice to have return the fully qualified type of the `Component`.

**Comment by cjhowe** [comment]
> So I'm thinking of maybe having the Component trait include another function: fn id() -> String or something similar.

This seems reasonable to me. If we put this in a separate trait, we could have a `_derive` crate to provide something like `#[derive(Id)]` which would be generated from the component struct name.

**Comment by cjhowe** [comment]
Maybe something like [this](https://github.com/cjhowe/named_type)?

**Comment by cjhowe** [comment]
I think deserializing a world with serde would be a challenge. Because the world needs components to be registered for deserialization to work, you need to create the world before deserializing it, but the [deserialize trait](https://docs.serde.rs/serde/trait.Deserialize.html) needs to produce an owned value out of nothing. A trait such as the one suggested by @Aceeri would be ideal, but I'm not sure how to make this compatible with serde.

**Comment by cjhowe** [comment]
For `DeserializeSeed` convenience functions for `serde` backends, see https://github.com/serde-rs/serde/issues/644, although `DeserializeSeed` still works today if you instantiate the `Deserializer` yourself.

**Comment by nchashch** [comment]
TL;DR: We need a way to serialize and deserialize components, which depend on an outside resource (like `AssetManager`/`gfx::Factory`) for creation.

For a component, serialized and deserialized forms of which correspond one to one, like:
```yaml
Velocity:
  x: 1.0
  y: 2.0
```

Serialization or deserialization can be done in one step:

serialized component | `serde` -> deserialized component
deserialized component | `serde` -> serialized component

But there are important kinds of components for which, serialized and deserialized forms don't correspond. For example `Renderable` component in amethyst contains gfx-rs handlers, which can't be serialized or deserialized directly.

A serialized version of `Renderable` would contain filenames of mesh and texture files, like this:
```yaml
Renderable:
  mesh: "sphere.obj"
  ambient: "ambient.dds"
  diffuse: "diffuse.dds"
  specular: "specular.dds"
```

So additional steps would be required in order to get from serialized `Renderable` to deserialized `Renderable`:

serialized `Renderable` |- `serde` -> Asset filenames |- `AssetManager` -> Assets in RAM |- `gfx::Factory` -> gfx-rs handlers |- -> deserialized `Renderable`

Maybe we could store `Option`s of gfx-rs handlers in `Renderable` and all the filenames for the assets. Then a `System` would:
1. walk all `Renderable`s
2. give jobs to `AssetManager` to load all the required assets
3. check every frame whether or not assets were loaded
4. if they were loaded, fill `Renderable`s with `Some(gfx-rs handler)`s.

Since filenames would probably be stored in deserialized version serialization would remain one step:

deserialized Renderable | `serde` -> serialized Renderable

**Comment by Aceeri** [comment]
@nchashch I think that could just be done with custom implementations of Serialize/Deserialize for those types of components. The thing I worry about is how you would get a reference to the asset manager.

**Comment by torkleyy** [comment]
I guess we can close this now. @kvark

**Comment by kvark** [comment]
Hooray!
